### ISSUE TITLE
[Feature Request] Support primary key (was: Use a succinct insert mechanism)

### ISSUE THREAD

**Issue by creativepsyco** [realm/realm-java#485]
Hi,

Currently all inserts happen like this:
```java
Realm realm = Realm.getInstance(context, dbName);
realm.beginTransaction();
MyObject object = realm.createObject(MyObject.class);
object.setName("John Doe");
object.setEmail("john@example.org");
realm.commitTransaction(); // Notification of update happens here
```

Ideally assuming we have tons of db objects (like nested classes) it would very wasteful to create an object (or retrieve an object) to update its fields. A shorter way would be to expose another function `realm.saveObject(MyObject.class, object)` which can update the db object if it exists. However, this would require the primary key implementation (or generated hashIds) which I think is currently not supported. 

Thoughts?

**Comment by cmelchior** [comment]
Standalone objects will be release as part of 0.77.0. This means that it should be possible to finish this now:

Features 
- Single primary key supported for now using @PrimaryKey annotation
- add ```createOrUpdateFromJson(Object)```
- add ```createOrUpdateFromJson(List)```
- add ```copyToRealmOrUpdate(Object)``` // Horrible name, anything better?
- add ```copyToRealmOrUpdate(List)``` // Horrible name, anything better?

Future work:
- Support for composite keys.
- Move primary key support to core.

**Comment by yolapop** [comment]
is @PrimaryKey ready? I can't seem to find it on 0.77.0 release.

**Comment by cmelchior** [comment]
@yolapop 
No, they have not been merged yet. We are planning for a 0.79 release, but might have to wait for 0.80 as we finishing up some other loose ends atm.

**Comment by Navideck-fotis** [comment]
Any updates on composite primary keys?

**Comment by cmelchior** [comment]
Composite keys will not be part of the initial version of Primary keys we plan to release. However we do plan to add it at a later stage. As a work around until then you can emulate composite keys by making a composite field of your own and mark it as a primary key.

**Comment by vvictor10** [comment]
@cmelchior Just to confirm, we can create a Class which holds multiple members and make that the primary key member on another entity class?

Also, is there an ETA on when primary keys could be available?

**Comment by cmelchior** [comment]
No, Primary keys will in its initial version only work on shorts, int, longs and Strings. Making them work on other RealmObjects is an interesting idea, and I can't see why it shouldn't be possible, so it is definitely something we will consider when we expand support for Primary keys.

Primary keys has been merged into master, we are just reviewing the associated createOrUpdate methods(), so soon :)

**Comment by Navideck-fotis** [comment]
I don't get it. Then how can I do what you suggested: "As a work around until then you can emulate composite keys by making a composite field of your own and mark it as a primary key."

What do you mean by composite field?

**Comment by cmelchior** [comment]
Eg. if your Primary key is the ```firstName``` and ```lastName``` you create a field called ```fullName``` which you then update with both values whenever you update one of them.

**Comment by vvictor10** [comment]
@cmelchior  In the real world, the composite primary key would be required not for fields like lastName and firstName, but more for stateName(String) and userId(Int), which would help in querying for users in a particular state. Just a random example, but you get the point.

**Comment by cmelchior** [comment]
Primary keys and createOrUpdate methods has been merged to master and will be part of 0.79.0

**Comment by vvictor10** [comment]
YES!! Thats awesome news @cmelchior . Any idea when 0.79 is slated to be released...
